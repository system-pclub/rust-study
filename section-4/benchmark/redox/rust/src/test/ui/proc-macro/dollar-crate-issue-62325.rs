// check-pass
// edition:2018
// aux-build:test-macros.rs
// aux-build:dollar-crate-external.rs

// Anonymize unstable non-dummy spans while still showing dummy spans `0..0`.
// normalize-stdout-test "bytes\([^0]\w*\.\.(\w+)\)" -> "bytes(LO..$1)"
// normalize-stdout-test "bytes\((\w+)\.\.[^0]\w*\)" -> "bytes($1..HI)"

#[macro_use]
extern crate test_macros;
extern crate dollar_crate_external;

type S = u8;

macro_rules! m { () => {
    #[print_attr]
    struct A(identity!($crate::S));
}}

m!();

dollar_crate_external::issue_62325!();

fn main() {}
