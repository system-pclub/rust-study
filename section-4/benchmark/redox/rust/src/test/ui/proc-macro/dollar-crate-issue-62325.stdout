PRINT-ATTR INPUT (DISPLAY): struct A(identity!(crate :: S));
PRINT-ATTR RE-COLLECTED (DISPLAY): struct A (identity ! ($crate :: S)) ;
PRINT-ATTR INPUT (DEBUG): TokenStream [
    Ident {
        ident: "struct",
        span: #3 bytes(LO..HI),
    },
    Ident {
        ident: "A",
        span: #3 bytes(LO..HI),
    },
    Group {
        delimiter: Parenthesis,
        stream: TokenStream [
            Ident {
                ident: "identity",
                span: #3 bytes(LO..HI),
            },
            Punct {
                ch: '!',
                spacing: Alone,
                span: #3 bytes(LO..HI),
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [
                    Ident {
                        ident: "$crate",
                        span: #3 bytes(LO..HI),
                    },
                    Punct {
                        ch: ':',
                        spacing: Joint,
                        span: #3 bytes(LO..HI),
                    },
                    Punct {
                        ch: ':',
                        spacing: Alone,
                        span: #3 bytes(LO..HI),
                    },
                    Ident {
                        ident: "S",
                        span: #3 bytes(LO..HI),
                    },
                ],
                span: #3 bytes(LO..HI),
            },
        ],
        span: #3 bytes(LO..HI),
    },
    Punct {
        ch: ';',
        spacing: Alone,
        span: #3 bytes(LO..HI),
    },
]
PRINT-ATTR INPUT (DISPLAY): struct B(identity!(::dollar_crate_external :: S));
PRINT-ATTR RE-COLLECTED (DISPLAY): struct B (identity ! ($crate :: S)) ;
PRINT-ATTR INPUT (DEBUG): TokenStream [
    Ident {
        ident: "struct",
        span: #10 bytes(LO..HI),
    },
    Ident {
        ident: "B",
        span: #10 bytes(LO..HI),
    },
    Group {
        delimiter: Parenthesis,
        stream: TokenStream [
            Ident {
                ident: "identity",
                span: #10 bytes(LO..HI),
            },
            Punct {
                ch: '!',
                spacing: Alone,
                span: #10 bytes(LO..HI),
            },
            Group {
                delimiter: Parenthesis,
                stream: TokenStream [
                    Ident {
                        ident: "$crate",
                        span: #10 bytes(LO..HI),
                    },
                    Punct {
                        ch: ':',
                        spacing: Joint,
                        span: #10 bytes(LO..HI),
                    },
                    Punct {
                        ch: ':',
                        spacing: Alone,
                        span: #10 bytes(LO..HI),
                    },
                    Ident {
                        ident: "S",
                        span: #10 bytes(LO..HI),
                    },
                ],
                span: #10 bytes(LO..HI),
            },
        ],
        span: #10 bytes(LO..HI),
    },
    Punct {
        ch: ';',
        spacing: Alone,
        span: #10 bytes(LO..HI),
    },
]
